File: src/web/js/org/zkace/Acebox.js
// Copyright (c) 2014 Darren Scott - All Rights Reserved
//
// This program is distributed under LGPL Version 2.1 in the hope that
// it will be useful, but WITHOUT ANY WARRANTY.
org.zkace.Acebox = zk.$extends(zul.inp.Textbox, {

	_editor : null,

	$define: {
		theme: null,
		mode: null,
		showgutter: null,
		maxrows: null,
		rowsSet: null
	},
  
	setValue: function(value, fromServer) {
		this.$supers('setValue', arguments);
		if (fromServer && this._editor) {
			this._editor.setValue(value);
		}
    },

	setHeight: function(value) {
		this.$supers('setHeight', arguments);
		if (this._editor) {
			this._editor.resize();
		}
	},

	setWidth: function(value) {
		this.$supers('setWidth', arguments);
		if (this._editor) {
			this._editor.resize();
		}
	},
	
	bind_: function() {
		this.$supers('bind_', arguments);
		var elem = this.$n(), widget = this, editor = ace.edit(this.uuid), session, base, lines, min, max;
		lines = editor.getSession().getDocument().getLength();
		base = zk.ajaxURI('/web/js/org/zkace/ace/',{au:true});
		ace.config.set('basePath', base);
		session = editor.getSession();
		this._editor = editor;
		if (this._theme) {
			editor.setTheme('ace/theme/' + this._theme);
		}
		if (this._mode) {
			session.setMode('ace/mode/' + this._mode);
		}
		editor.setReadOnly(this._readonly);
		editor.renderer.setShowGutter(this._showgutter);
		min = this.getRowsSet() ? this.getRows() : Math.min(lines, 1);
		max = this.getMaxrows() >= min ? this.getMaxrows() : min;
		editor.setOptions({minLines: min, maxLines: max});
		// TODO Make this a component property
		// editor.setBehavioursEnabled(true);
		session.setUseWrapMode(true);
		session.setWrapLimitRange();
		editor.on('blur', function(e) {
			widget.setValue(editor.getValue());
			widget.fireOnChange();
		});
		session.on("changeAnnotation", function(){
			var annot = session.getAnnotations();
			widget.smartUpdate('annotations', annot);
		});
	},

	redraw: function(out) {
		// out.push('<div', this.domAttrs_(), '>');
		out.push('<div', this.domAttrs_(), '>');
		if (this.getValue()) {
			out.push(this.getValue().replace(/&/g, '&amp').replace(/</g, '&lt;'));
		}
		out.push('</div>');
	}
        
});

File: src/org/zkace/Acebox.java
/*
 * Copyright (c) 2014 Darren Scott - All Rights Reserved
 * 
 * This program is distributed under LGPL Version 2.1 in the hope that
 * it will be useful, but WITHOUT ANY WARRANTY.
 */
package org.zkace;

import java.io.IOException;

import org.zkoss.zk.ui.WrongValueException;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.Textbox;

public class Acebox extends Textbox {

	private static final long serialVersionUID = 1L;

	private boolean rowsSet = false;
	
	private boolean showgutter = false;

	private int maxrows = -1;
	
	private String mode;
	
	private String theme;

	public Acebox() {
		super();
		this.setMultiline(true);
	}

	@Override
	protected void renderProperties(ContentRenderer renderer) throws IOException {
		super.renderProperties(renderer);
		render(renderer, "mode", this.mode); 
		render(renderer, "theme", this.theme); 
		render(renderer, "showgutter", this.showgutter); 
		render(renderer, "rowsSet", this.rowsSet); 
		render(renderer, "maxrows", this.maxrows); 
	}
	
	@Override
	public void setRows(int rows) throws WrongValueException {
		super.setRows(rows);
		this.rowsSet = true;
		this.smartUpdate("rowsSet", true, true);
	}
	
	public void setMode(String mode) {
		this.mode = mode;
		this.smartUpdate("mode", mode, true);
	}
	
	public String getMode() {
		return mode;
	}

	public void setTheme(String theme) {
		this.theme = theme;
		this.smartUpdate("theme", theme, true);
	}
	
	public String getTheme() {
		return theme;
	}
	
	public void setShowgutter(boolean showgutter) {
		this.showgutter = showgutter;
		this.smartUpdate("showgutter", showgutter, true);
	}
	
	public boolean getShowgutter() {
		return showgutter;
	}

	public void setMaxrows(int maxrows) {
		this.maxrows = maxrows;
		this.smartUpdate("maxrows", maxrows, true);
	}
	
	public int getMaxrows() {
		return maxrows;
	}

}